endpoint: Do not override deny entries with proxy redirects

jrajahalme opened the pull request:
Use DenyPreferredInsert instead of directly manipulating policy map state to make sure deny entries are not overridden by new proxy redirect entries.

Prior to this fix it was possible for a proxy redirect to be pushed onto the policy map when it should have been overridden by a deny at least in these cases:
- L3-only deny with L3/L4 redirect: No redirect should be added as the L3 is denied, the new test case verifies this
- L3-only deny with L4-only redirect: L4-only redirect should be added and an L3/L4 deny should also be added, but the L3/L4 deny is only added by deny preferred insert, and is missed when the map is manipulated directly.

It is clear that in the latter case the addition of the redirect can not be completely blocked, so we can't fix this by making AllowsL4 more restrictive. But also in the former case it is possible that the deny rule only covers a subset of security identities, while the redirect rule covers some of the same security identities, but also some more that should not be blocked. Hence the correct fix here is to leave `AllowsL4` to be L3-independent, and cover these cases with deny preferred insert instead of adding redirect entries to the map directly.

This PR also contains a related change that allows a redirect entry to be updated, maybe with a changed proxy port. I've not seen evidence that this is currently fixing a bug, but it feels like a real possibility.

Fixes: #12716

```release-note
Fixed proxy redirect policy implementation when any deny rule prevents them.
```

jrajahalme committed:
endpoint: Do not override deny entries with proxy redirects

Use DenyPreferredInsert instead of directly manipulating policy map state
to make sure deny entries are not overridden by new proxy redirect
entries.

Prior to this fix it was possible for a proxy redirect to be pushed onto
the policy map when it should have been overridden by a deny at least in
these cases:

- L3-only deny with L3/L4 redirect: No redirect should be added as the L3
  is denied

- L3-only deny with L4-only redirect: L4-only redirect should be added
  and an L3/L4 deny should also be added, but the L3/L4 deny is only added
  by deny preferred insert, and is missed when the map is manipulated
  directly. A new test case verifies this.

It is clear that in the latter case the addition of the redirect can not
be completely blocked, so we can't fix this by making AllowsL4 more
restrictive. But also in the former case it is possible that the deny
rule only covers a subset of security identities, while the redirect rule
covers some of the same security identities, but also some more that
should not be blocked. Hence the correct fix here is to leave AllowsL4 to
be L3-independent, and cover these cases with deny preferred insert
instead of adding redirect entries to the map directly.

This commit also contains a related change that allows a redirect entry
to be updated, maybe with a changed proxy port. I've not seen evidence
that this is currently fixing a bug, but it feels like a real
possibility.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

christarazi commented on pkg/endpoint/bpf.go:
```diff
@@ -302,22 +302,20 @@ func (e *Endpoint) addNewRedirectsFromDesiredPolicy(ingress bool, desiredRedirec
 					entry.ProxyPort = redirectPort
 				}
-				e.desiredPolicy.PolicyMapState[keyFromFilter] = entry
+				e.desiredPolicy.PolicyMapState.DenyPreferredInsertWithChanges(keyFromFilter, entry, insertedDesiredMapState, nil, idLookup)
```
Is it expected that `DenyPreferredInsertWithChanges` will insert into `insertedDesiredMapState`?

jrajahalme replied:
Previously we were inserting into `insertedDesiredMapState` only when the key/entry did not exist previously, now `DenyPreferredInsertWithChanges` inserts into `insertedDesiredMapState` also when the there was a previous entry with the same key. This does not matter, as where these are used in the revert code below, we first delete all inserted keys, then add back the old values, if they existed, from `updatedDesiredMapState`. End result is that same before and after, previously the revert code directly replaced the new value with the old one, now the new value is deleted first, and the old one added back after.

jrajahalme replied:
Hmm. if there were old values also for (e.g., L3L4) entries created by `DenyPreferredInsertWithChanges` then the old values of those can not be reverted back in case of a failure with the current code in the PR. I may have to change `adds` and `deletes` arguments to `DenyPreferredInsertWithChanges` to be `MapState` instead of just `Keys` to make `DenyPreferredInsertWithChanges` revertible. Let me try that!

christarazi commented on pkg/policy/mapstate.go:
```diff
@@ -504,12 +504,17 @@ func (keys MapState) denyPreferredInsertWithChanges(newKey Key, newEntry MapStat
 	// Merging owners from the new entry to the existing one has no datapath impact so we skip
 	// adding anything to 'adds' here.
-	if oldEntry, exists := keys[key]; exists && (oldEntry.IsRedirectEntry() || oldEntry.IsDeny) {
+	// Merging owners has the effect of keeping the shered entry alive as long as one of the owners exists.
```
typo: 
```suggestion
	// Merging owners has the effect of keeping the shared entry alive as long as one of the owners exists.
```

jrajahalme replied:
Fixed, thanks!

christarazi commented on pkg/policy/mapstate.go:
```diff
@@ -504,12 +504,17 @@ func (keys MapState) denyPreferredInsertWithChanges(newKey Key, newEntry MapStat
-	if oldEntry, exists := keys[key]; exists && (oldEntry.IsRedirectEntry() || oldEntry.IsDeny) {
+	// Merging owners has the effect of keeping the shered entry alive as long as one of the owners exists.
+	if oldEntry, exists := keys[key]; exists &&
+		(!entry.IsRedirectEntry() && oldEntry.IsRedirectEntry() || oldEntry.IsDeny) {
```
IIUC, this is not changing functional behavior and we are just adding an extra condition for explicitness for the if-statement, correct?

jrajahalme replied:
Right now I think that is the case, yes. I was thinking that if the proxy port of an existing entry needs to be updated, then this code path should allow for that. I *think* this is currently not happening, but we are replacing direct manipulation of the redirect entry in `bpf.go` with the `DenyPreferredInsertWithChanges` that will end up here for a non-deny entry, so I thought it would be best to allow for proxy port to be updated by that call, if it ever needs to happen.

jrajahalme replied:
Updated with a new 1st commit that exports `DenyPreferredInsertWithChanges()` and makes it revertible, including unit testing.

joestringer commented on pkg/endpoint/bpf.go:
It seems to me like this line was intended to try to protect against this case previously, but it relies on the desired `mapstate` while iterating the L4 filters, which I guess is insufficient in this case for some reason? Do we still need this check? I find it a bit confusing that the reader cannot make assumptions about whether the traffic is allowed after this line:

https://github.com/cilium/cilium/blob/3fc1e8798ff98dd1385cc884284ad0f2fc671895/pkg/endpoint/bpf.go#L230 

Walking back up the stack to `runPreCompilationSteps()`, it _seems_ to me that the desired policy map should be populated with the content to be able to allow the above check to protect against this condition already.

joestringer replied:
Ah, I see that this patch now adds the comment inside this function to state:

```
// Note: Traffic to/from a specific L3 on any of the selectors could still be denied!
```

but I don't quite understand, is it not possible to make a clear call about whether this L4 filter is allowed using the map state? Is that because the L4 filter may contain a range of L3 peers or even the "any" L3 peer (ie, L4-only rule/filter), and as such we can't make a generic deny query into the mapstate without iterating the L4 filter's L3 peers and resolving those against a selectorcache for the real identities?

joestringer replied:
At best it is definitely misleading to see `!mapState.AllowsL4(...)` and it turns out that actually the mapState may still disallow the L4 peer later on somehow due to L3-only policies :thinking: ...I guess that's the origin of how we missed this bug previously.

jrajahalme replied:
I think `AllowsL4()` has value in avoiding creating redirects when they are "clearly" denied (all traffic is denied, or all traffic on the port of the redirect is denies). I guess `L4` in the function name may refer to the fact that it is not checking for `L3`. In general, it is not possible deduce that a redirect should not be created based on the current contents of the desired map state, as a FQDN policy may only later produce the IDs that are allowed, i.e., at the time of `addRedirects()` the identity that will actually be allowed to have a redirect may not exist yet, it is not possible to check for that yet-non-existing key value.

It might be better to invert the return value and the name of `AllowsL4` to `DeniesL4`, returning true if the given L4 is denied for everybody?

jrajahalme replied:
Did the rename.

joestringer commented on pkg/policy/mapstate.go:
```diff
@@ -498,24 +560,28 @@ func (keys MapState) denyPreferredInsertWithChanges(newKey Key, newEntry MapStat
+	// Merging owners has the effect of keeping the shared entry alive as long as one of the owners exists.
+	if oldEntry, exists := keys[key]; exists &&
+		(!entry.IsRedirectEntry() && oldEntry.IsRedirectEntry() || oldEntry.IsDeny) {
 		oldEntry.MergeReferences(&entry)
 		keys[key] = oldEntry
 		return
```
This case doesn't leave an entry in `old`. Could that lead to a reference leak?

jrajahalme replied:
Good catch, added!

joestringer reviewed (approved):
One minor referencing question below.

jrajahalme committed:
policy: Export DenyPreferredInsertWithChanges, make revertible

Export DenyPreferredInsertWithChanges and make it revertible by taking a
map of old values as a new optional argument.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>
